### ISSUE TITLE
CI not completing

### ISSUE THREAD

**Issue by filipefurtad0** [openfoodfoundation/openfoodnetwork#10192]
## Description

The CI hangs and does not complete.

## Steps to Reproduce

1. Create a PR and watch the CI hanging.

## Animated Gif/Screenshot

Knapsack dashboard - builds from `knapsack_rspec_system `
https://knapsackpro.com/dashboard/organizations/3491/projects/2077/test_suites/3042/builds

![image](https://user-images.githubusercontent.com/49817236/209118713-65bead39-7671-4c83-a40f-d192300d6600.png)

**Comment by filipefurtad0** [comment]
If we look at the output of this build run
https://github.com/openfoodfoundation/openfoodnetwork/actions/runs/3756398583/jobs/6382433093

it mentions
`I , [2022-12-22T09:10:30.421868 #3075]  INFO -- : [knapsack_pro] bundle exec rspec --format progress --default-path spec "spec/system/admin/product_import_spec.rb"
`

I've changed this spec yesterday:
https://github.com/openfoodfoundation/openfoodnetwork/pull/10188/commits/02ab4bfe305b66f4bf4d13862f3bbeec9be448a8

Coincidence?
Investigating.

**Comment by filipefurtad0** [comment]
My first guess was that a binding.pry was introduced in the CI, as here:
https://github.com/openfoodfoundation/openfoodnetwork/pull/10188/commits/02ab4bfe305b66f4bf4d13862f3bbeec9be448a8

![image](https://user-images.githubusercontent.com/49817236/209121342-7ce4f70c-9713-479a-a8e1-6d1e25d69cbd.png)

It is commented out, so that should not be the issue here.

**Comment by filipefurtad0** [comment]
We could be looking at this issue:
https://knapsackpro.com/faq/question/why-knapsack_pro-hangs-freezes-is-stale-for-tests-in-chrome-disallow-downloads-in-sandboxed-iframes

Or this issue:
https://knapsackpro.com/faq/question/why-knapsack_pro-freezes--hangs-my-ci-for-instance-travis

Not sure why these surfaced all of the sudden...

**Comment by filipefurtad0** [comment]
I can't seem to run specs on latest master, also locally:

![image](https://user-images.githubusercontent.com/49817236/209196517-58f7bc53-38ad-4cdc-a685-e4f28cf9b158.png)

**Comment by jibees** [comment]
I'm up-to-date with upstream master, and I can run `bundle exec rspec spec/system/admin/users_spec.rb` (everything is green).

**Comment by jibees** [comment]
Not always the same that takes long time. Sometime it's `knapsack_rspec_system_admin (10, 4)`, or `knapsack_rspec_system_admin (10, 3)` or `knapsack_rspec_system_admin (10, 6)`, ...

**Comment by jibees** [comment]
@filipefurtad0 If you can't run the spec locally, I think it's a good way to continue to investigate. For example, maybe we need to understand why I can run the spec locally.

**Comment by filipefurtad0** [comment]
It seems not only system specs:
![image](https://user-images.githubusercontent.com/49817236/209339744-755f0726-f16d-4302-9314-7c6826167bc3.png)

**Comment by filipefurtad0** [comment]
> I'm up-to-date with upstream master, and I can run bundle exec rspec spec/system/admin/users_spec.rb (everything is green).

Worked for me as well today:

```

 Precompiling assets.
Capybara starting Puma...
* Version 5.6.5 , codename: Birdie's Version
* Min threads: 0, max threads: 4
* Listening on http://127.0.0.1:44825
...........

Finished in 32.4 seconds (files took 7.56 seconds to load)
11 examples, 0 failures
```

Tricky...

**Comment by filipefurtad0** [comment]
Ah. Got it:

![Peek 2022-12-23 15-04](https://user-images.githubusercontent.com/49817236/209356994-b7e084b1-1a49-47e7-ad7a-c4fe09dbf46e.gif)

**Comment by filipefurtad0** [comment]
Steps to reproduce:
`bundle exec rspec spec/system/admin/order_cycles/list_spec.rb:151`

This will run endlessly, after 2022.12.22 - which was _yesterday_.... It seems the future caught up with the past!! :scream: 

![image](https://user-images.githubusercontent.com/49817236/209358818-dc340eb4-ae6f-46db-b137-acc4b11c9705.png)

**Comment by RachL** [comment]
What a way to end the year :joy:

**Comment by filipefurtad0** [comment]
Happening again:

![image](https://user-images.githubusercontent.com/49817236/213151066-89ea9fa4-117f-4e94-b5ca-cab633a184d8.png)

**Comment by filipefurtad0** [comment]
Looking at this issue/build run:
https://github.com/openfoodfoundation/openfoodnetwork/actions/runs/3948056612/jobs/6757604661

And clicking the gear icon => "View raw logs", we can see that the testing did not really start:

```

2023-01-18T10:41:15.8198924Z Requested labels: ubuntu-20.04
2023-01-18T10:41:15.8198969Z Job defined at: openfoodfoundation/openfoodnetwork/.github/workflows/build.yml@refs/pull/10249/merge
2023-01-18T10:41:15.8198997Z Waiting for a runner to pick up this job...
2023-01-18T10:41:25.2060171Z Job is waiting for a hosted runner to come online.
2023-01-18T10:41:25.4106867Z Job is about to start running on the hosted runner: GitHub Actions 31 (hosted)
```

**Comment by filipefurtad0** [comment]
Ok, got it - this selector keeps increasing the dates, searching for a date which will not be found:

![image](https://user-images.githubusercontent.com/49817236/213168239-69510981-d1f8-43f3-b400-fc1255a2c56d.png)

Steps to reproduce:
- run `bundle exec rspec -fd "sec/system/admin/bulk_order_management_spec.rb:488`

**Comment by filipefurtad0** [comment]
This is tough: the issue is related to the fact that the start date of the selected time interval, is not visible:

![image](https://user-images.githubusercontent.com/49817236/213177370-e6886516-e485-4d13-9d97-2bf6bce13de0.png)

I wonder how we could improve our helper, so that it fetches dates from previous months too. Currently, we formulate this in our `select_date_from_datepicker` like so:
```
    def select_date_from_datepicker(date, reference_date = Time.zone.today)
      navigate_datepicker_to_month(date, reference_date)
      find('.flatpickr-calendar.open .flatpickr-days .flatpickr-day:not(.prevMonthDay)',
           text: date.strftime("%e").to_s.strip, exact_text: true, match: :first).click
    end
```

So either we change the helper, or we make sure we always select dates fitting within our calendar:
![image](https://user-images.githubusercontent.com/49817236/213176084-f68dca2f-0fb6-4e88-9a47-9f313907c605.png)

I'm not sure what the best option is... What do you think @jibees ?

**Comment by jibees** [comment]
Still an issue with the date selector: good catch @filipefurtad0 !

Locally, running: 
```
bundle exec rspec -fd spec/system/admin/bulk_order_management_spec.rb:488
```

creates a failed spec actually:

```
Failure/Error: expect(find("input.datepicker").value).to eq "#{oc3.orders_open_at.strftime('%F')} to #{oc3.orders_close_at.strftime('%F')}"
     
       expected: "2023-01-26 to 2023-02-02"
            got: "2023-01-25 to 2023-02-01"
     
       (compared using ==)
```

**Comment by filipefurtad0** [comment]
> EDIT: if I comment the line #490, the test pass; this line will be deleted with https://github.com/openfoodfoundation/openfoodnetwork/pull/10243

Yes, agree. I'm closing this issue, as the CI is completing. Let's re-open if something like this occurs again :+1: 

**Lessons learned**

I guess datetime pickers and time-zones are the main suspects, if this happens again. For future debugging, maybe good to keep in mind of an option to debug, based in [this documentation](https://knapsackpro.com/faq/question/how-to-run-tests-for-particular-ci-node-in-your-development-environment):
- on your local environment, checkout to a failing branch and pull the last commit
- login to the Knapsack dashboard and using the commit hash, find the failing/unfinished node
- use the nr. of nodes and the nr. of the failing/unfinished node to execute a dry run of the knapsack routine:
`KNAPSACK_PRO_TEST_SUITE_TOKEN_RSPEC=ff2456e64c9f2aa5157eb0daf711d3c3 KNAPSACK_PRO_REPOSITORY_ADAPTER=git KNAPSACK_PRO_PROJECT_DIR=~/openfoodnetwork/ KNAPSACK_PRO_CI_NODE_TOTAL=10 KNAPSACK_PRO_CI_NODE_INDEX=8 KNAPSACK_PRO_TEST_FILE_PATTERN="{spec/system/admin/**/*_spec.rb}" bundle exec rake "knapsack_pro:queue:rspec[--dry-run]"`

On the syntax above, we'll need to use the correct token, and the queue option (for admin and consumer specs only).

This should output a list of specs. Within these, the culprit should be found. If we don't really know which one could it be then we'll need to run them all; but sometimes we can make a good "guess". Here for example, we'd be looking at specs which use the datepicker.

**Comment by filipefurtad0** [comment]
Hello again :wave: 
https://github.com/openfoodfoundation/openfoodnetwork/actions/runs/3998473950/jobs/6861164706

**Comment by filipefurtad0** [comment]
Ah! I guess it took so long because it kept repeating several tests which triggered:
```

            Ferrum::DeadBrowserError:
              Browser is dead or given window is closed
``` 

So, those tests x 60sec timout time looked like it was not completing... Closing again, false alarm :smile:
